images/devtools-golang-v1beta1/context/Makefile

```diff
@@ -54,146 +58,234 @@ endef
+
 .PHONY: test
-test:  ## Dynamic code validation
+test: ## Run the project's tests.
+
+.PHONY: validate
+validate: ## Validate everything.
+validate: validate-static test
+
+.PHONY: build
+build: ## Build the source.
+
+.PHONY: watch
+watch: ## Watch code for changes and re-validate.
+
+.PHONY: install
+install: ## Install globally.
+
 .PHONY: publish
-publish: ## publish artifacts
+publish: ## Publish artifacts.
+
+.PHONY: deploy
+deploy: ## Deploy the project.
+
+.PHONY: all
+all: ## Do everything.
+all: validate generate build
 
 ########################################################################
 # golang
 ########################################################################
 
-validate-fix: validate-golang-fix
-.PHONY: validate-golang-fix
-validate-golang-fix:
-	$(golangci_lint) run  --fix ./...
+golangci_lint=golangci-lint $(if $(filter all commands,$(VERBOSE)),-v)
 
+validate-fix: golang-validate-fix
+.PHONY: golang-validate-fix
+golang-validate-fix:
+	$(golangci_lint) run  --fix ./...
 
-.PHONY: validate-golang-static
-validate-static: validate-golang-static
-validate-golang-static:
+.PHONY: golang-validate-static
+validate-static: golang-validate-static
+golang-validate-static:
 	$(golangci_lint) run  ./...
 
-.PHONY: validate-golang
-validate: validate-golang
-validate-golang: validate-golang-static test-golang  ## Run static and dynamic validation (tests)
-
-.PHONY: test-golang
-test: test-golang
-test-golang coverage.out:
+.PHONY: golang-test
+test: golang-test
+golang-test: ## Test golang code.
+golang-test coverage.out: | $(localstatedir)/
 	go test -cover \
 		-race \
 		$(go_build_flags) \
 		-coverprofile=coverage.out -covermode=atomic \
-		$(if $(filter all commands,$(VERBOSE)),-v) ./... \
-		$(if $(gotest_files),$(gotest_files),./...) \
-		$(gotest_args)
+		$(gotest_args) \
+		$(if $(filter all commands,$(VERBOSE)),-v) \
+		$(if $(gotest_files),$(gotest_files),$(or $(CLI_ARGS),./...)) \
 
-.PHONY: view-coverage
-view-coverage: coverage.out ## view coverage
+
+.PHONY: golang-validate
+validate: golang-validate
+golang-validate: golang-validate-static golang-test  ## Run static and dynamic validation (tests).
+
+.PHONY: golang-view-coverage
+golang-view-coverage: coverage.out ## View golang coverage.
 	go tool cover -html=$(<)
 
-.PHONY: build
-build: ## build source
+.PHONY: golang-build
+golang-build: ## Build golang source.
 	go build -race \
 		$(go_build_flags) \
 		$(go_ldflags) \
 		$(if $(filter all commands,$(VERBOSE)),-v) \
 		./...
 
-.PHONY: build-golang-outputs
-build: build-golang-outputs
-build-golang-outputs: ## build golang outputs
-build-golang-outputs: | $(outputs_dir)/
+.PHONY: golang-build-outputs
+build: golang-build-outputs
+golang-build-outputs: ## Build golang outputs.
+golang-build-outputs: | $(outputs_dir)/
 	CGO_ENABLED=0 go build \
 		$(go_build_flags) \
 		$(go_ldflags) \
 		$(if $(filter all commands,$(VERBOSE)),-v) \
 		-o $(outputs_dir) \
 		./...
 
+.PHONY: golang-generate
```

Comment:
If it's golang specific docker file, do we need to add `golang` to most of the commands? Wouldn't be `build-outputs` enough? If I used `build-outputs` in golang specific doker file I would expect to generate go outputs.

Reply:
There are quite a few tools interacting here already: golang, protobuf, buildx.

For golang though building is a bit odd, golang build does not generate outputs, it builds and then throws away the output, unless you specify a directory.

This however is more in line with doing a normal build, so golang-build-outputs is aliased to `build`.

So normally, you just run the `build` target, and that does what you probably want, which is creating the final outputs.

However, it will also build the docker image. If there is just something broken with building golang, and the docker image builds fine, then you will likely just want to run the golang build, to debug that, which is why we have toolchain specific targets.


In the case of generate you should in most cases just call `generate` and then all code generation should happen. I do link toolchain specific targets to generic targets, e.g.:

```makefile
.PHONY: golang-generate
generate: golang-generate
golang-generate: ## Generate golang things.
golang-generate:
```

So generate should call golang-generate.

But there is also buf-generate, which only gets linked in when you enable protobuf. And again maybe one of the two is broken, and then to facilitate debugging you want to just run the specific thing causing problems.

Reply:
Possibly I should rename `golang-build-outputs` to `golang-build` and `golang-build` to `golang-build-null` or something.

I will think about it, but in general developers should stick to the non toolchain specific targets as much as possible, so you really should just do `validate` and `build`. I do that for the most part.